libtenzir/builtins/functions/community_id.cpp

```diff
@@ -0,0 +1,180 @@
+      if (not dst_ips) {
+        diagnostic::warning("`community_id` expected `ip` as 3rd argument")
+          .primary(args.dst_ip)
+          .emit(ctx);
+        return null_series();
+      }
+      auto dst_ports = dst_port_series.as<int64_type>();
+      if (not dst_ports) {
+        diagnostic::warning("`community_id` expected `integer` as 4th argument")
+          .primary(args.dst_port)
+          .emit(ctx);
+        return null_series();
+      }
+      auto protos = proto_series.as<string_type>();
+      if (not protos) {
+        diagnostic::warning("`community_id` expected `string` as 5th argument")
+          .primary(args.dst_port)
+          .emit(ctx);
+        return null_series();
+      }
+      auto seeds = std::optional<basic_series<int64_type>>{};
+      if (seed_series) {
+        seeds = seed_series->as<int64_type>();
+        if (not seeds) {
+          diagnostic::warning("`community_id` got an argument type mismatch")
+            .primary(*args.seed)
+            .note("expected argument `seed` to be of type `int64`")
+            .emit(ctx);
+        }
+      }
+      auto b = arrow::StringBuilder{};
+      check(b.Reserve(eval.length()));
+      for (auto i = int64_t{0}; i < eval.length(); ++i) {
+        if (src_ips->array->IsNull(i) or src_ports->array->IsNull(i)
+            or dst_ips->array->IsNull(i) or dst_ports->array->IsNull(i)
+            or protos->array->IsNull(i)) {
+          check(b.AppendNull());
+          continue;
+        }
+        const auto* src_ip_ptr = src_ips->array->storage()->GetValue(i);
+        const auto* dst_ip_ptr = dst_ips->array->storage()->GetValue(i);
+        auto src_ip = ip::v6(as_bytes<16>(src_ip_ptr));
+        auto dst_ip = ip::v6(as_bytes<16>(dst_ip_ptr));
+        auto src_port = src_ports->array->GetView(i);
+        auto dst_port = dst_ports->array->GetView(i);
+        auto proto = protos->array->GetView(i);
+        auto type = port_type::unknown;
+        if (proto == "tcp") {
+          type = port_type::tcp;
+        } else if (proto == "udp") {
+          type = port_type::udp;
+        } else if (proto == "icmp") {
+          type = port_type::icmp;
+        } else if (proto == "icmp6") {
+          type = port_type::icmp6;
+        } else {
+          diagnostic::warning("`community_id` expected `tcp`, `udp`, `icmp`, "
+                              "or `icmp6` as protocol")
+            .primary(args.proto)
+            .emit(ctx);
+          check(b.AppendNull());
+          continue;
+        }
+        auto flow
+          = make_flow(src_ip, dst_ip, detail::narrow<uint16_t>(src_port),
+                      detail::narrow<uint16_t>(dst_port), type);
+        auto seed = uint16_t{0};
+        if (seeds and not seeds->array->IsNull(i)) {
+          // TODO: Perform a bounds check. There are probably already utilities
+          // for this available.
+          seed = detail::narrow_cast<uint16_t>(seeds->array->GetView(i));
```

Comment:
@jachris here's your need for `try_narrow` again 😅

Reply:
Yeah, we should definitely add this. 

What should happen if the port is out of range? Add a `null`, and emit a warning (one per batch)?

Reply:
I've implemented a bounds check manually.